REGRESSION(284617@main): Some HDR images may be displayed as black boxes when converting to SDR

shallawa opened the pull request:
#### 78f467e29fbac5ce8a733d7676e542aaaf7fa348
<pre>
REGRESSION(284617@main): Some HDR images may be displayed as black boxes when converting to SDR
<a href="https://bugs.webkit.org/show_bug.cgi?id=282420">https://bugs.webkit.org/show_bug.cgi?id=282420</a>
<a href="https://rdar.apple.com/139041771">rdar://139041771</a>

Reviewed by Simon Fraser.

284617@main made NativeImage::draw() no longer convert the HDR image to SDR image
using the un-accelerated GraphicsContext. But ShareableBitmap needs to create its
GraphicsContext with extendedSRGBColorSpace to get the HDR image drawn correctly.

Currently DestinationColorSpace::asRGB() returns sRGBColorSpace true, if the image
colorspace usesRec2100TransferFunctions(). DestinationColorSpace::asRGB() should
return nullopt in this case. This will force validateColorSpace() to return
extendedSRGBColorSpace.

NativeImage::draw() converts the image to extendedSRGBColorSpace if its colorspace
usesRec2100TransferFunctions(). So convertToDestinationColorSpaceIfNeeded() can
be deleted.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::asRGB const):
(WebCore::DestinationColorSpace::usesExtendedRange const):
(WebCore::DestinationColorSpace::usesRec2100TransferFunctions const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
(WebCore::DestinationColorSpace::usesStandardRange const):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebCore::ShareableBitmapConfiguration::calculateBitmapInfo):
(WebCore::wantsExtendedRange): Deleted.
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::GraphicsContextCG::convertToDestinationColorSpaceIfNeeded): Deleted.

Canonical link: <a href="https://commits.webkit.org/286164@main">https://commits.webkit.org/286164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f8d21b067ed5a92c89f36e9a4a5a743b7754ac6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74957 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27774 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79425 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77074 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2172 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58875 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17124 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78024 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49013 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64397 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39262 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46319 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21896 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24528 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68094 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67458 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22240 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80873 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74215 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2275 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1375 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67108 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2424 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64416 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66408 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16517 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10347 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8507 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96486 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2240 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5028 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21086 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2268 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3189 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2275 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
```diff
@@ -140,6 +138,23 @@ bool DestinationColorSpace::supportsOutput() const
 #endif
 }
 
+bool DestinationColorSpace::isHDR() const
```
Doesn't seem right to conflate HDR and extended color range.  What is the actual data you are trying to determine here? Is it a specific transfer function? In general, I would love to avoid the term HDR in our color code as much as possible, as it lacks a clear agreed upon definition. Instead, let's try to be accurate with what exactly we are trying to check for.

shallawa replied:
Actually I want to check if the colorspace is `!isSDR()`. In this case we need to create a `ShareableBitmap` with `extendedSRGBColorSpace` to be able to transfer the image pixels to GPUProcess correctly.

smfr commented on Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
```diff
@@ -140,6 +138,26 @@ bool DestinationColorSpace::supportsOutput() const
 #endif
 }
 
+bool DestinationColorSpace::isXDR() const
```
Let's call this `isExtendedRange()`.

shallawa replied:
I renamed it `usesExtendedRange()` to be compatible with CG function name.

smfr commented on Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
```diff
@@ -140,6 +138,26 @@ bool DestinationColorSpace::supportsOutput() const
+#endif
+}
+
+bool DestinationColorSpace::isHDR() const
```
Rather than using `HDR`, maybe we should just talk about `ITUR_2100TF` perhaps with a nicer name.

shallawa replied:
I renamed it to `usesRec2100TransferFunctions()` to be compatible with CG function name.

smfr reviewed (approved):
Can we make a layout test for this, or does it require HDR display?

shallawa committed:
REGRESSION(284617@main): Some HDR images may be displayed as black boxes when converting to SDR
https://bugs.webkit.org/show_bug.cgi?id=282420
rdar://139041771

Reviewed by Simon Fraser.

284617@main made NativeImage::draw() no longer convert the HDR image to SDR image
using the un-accelerated GraphicsContext. But ShareableBitmap needs to create its
GraphicsContext with extendedSRGBColorSpace to get the HDR image drawn correctly.

Currently DestinationColorSpace::asRGB() returns sRGBColorSpace true, if the image
colorspace usesRec2100TransferFunctions(). DestinationColorSpace::asRGB() should
return nullopt in this case. This will force validateColorSpace() to return
extendedSRGBColorSpace.

NativeImage::draw() converts the image to extendedSRGBColorSpace if its colorspace
usesRec2100TransferFunctions(). So convertToDestinationColorSpaceIfNeeded() can
be deleted.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::asRGB const):
(WebCore::DestinationColorSpace::usesExtendedRange const):
(WebCore::DestinationColorSpace::usesRec2100TransferFunctions const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
(WebCore::DestinationColorSpace::usesStandardRange const):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebCore::ShareableBitmapConfiguration::calculateBitmapInfo):
(WebCore::wantsExtendedRange): Deleted.
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::GraphicsContextCG::convertToDestinationColorSpaceIfNeeded): Deleted.

Canonical link: https://commits.webkit.org/286164@main
